libweston/backend-rdp/rdp.c

```diff
@@ -370,182 +382,81 @@ ensure_matching_mode(struct weston_output *output, struct weston_mode *target)
 	return rdp_insert_new_mode(output, target->width, target->height, b->rdp_monitor_refresh_rate);
 }
 
-static int
-rdp_switch_mode(struct weston_output *output, struct weston_mode *target_mode)
+static void
+rdp_output_set_mode(struct weston_output *base, struct weston_mode *mode)
 {
-	struct rdp_output *rdpOutput = container_of(output, struct rdp_output, base);
-	struct rdp_backend *rdpBackend = to_rdp_backend(output->compositor);
-	struct rdp_peers_item *rdpPeer;
-	rdpSettings *settings;
 	pixman_image_t *new_shadow_buffer;
-	struct weston_mode *local_mode, *previous_mode;
+	struct rdp_output *rdpOutput = container_of(base, struct rdp_output, base);
+	struct rdp_backend *b = to_rdp_backend(base->compositor);
+	struct weston_mode *cur;
+	struct weston_output *output = base;
 	const struct pixman_renderer_output_options options = { .use_shadow = true, };
-	bool HiDefRemoteApp = false;
-
-	if (rdpBackend->rdp_peer && rdpBackend->rdp_peer->settings->HiDefRemoteApp)
-		HiDefRemoteApp = true;
-
-	local_mode = ensure_matching_mode(output, target_mode);
-	if (!local_mode) {
-		rdp_debug_error(rdpBackend, "mode %dx%d not available\n", target_mode->width, target_mode->height);
-		return -ENOENT;
-	}
-
-	if (local_mode == output->current_mode)
-		return 0;
+	struct rdp_peers_item *rdpPeer;
+	rdpSettings *settings;
 
-	if (HiDefRemoteApp)
-		previous_mode = output->current_mode;
-	else
-		output->current_mode->flags &= ~WL_OUTPUT_MODE_CURRENT;
+	mode->refresh = b->rdp_monitor_refresh_rate;
+	cur = ensure_matching_mode(base, mode);
+	cur->flags |= WL_OUTPUT_MODE_CURRENT;
+	cur->flags |= WL_OUTPUT_MODE_PREFERRED;
 
-	output->current_mode = local_mode;
-	output->current_mode->flags |= WL_OUTPUT_MODE_CURRENT;
+	if (base->current_mode)
+		base->current_mode->flags = 0;
 
-	if (HiDefRemoteApp) {
-		/* Mark current mode as preferred mode */
-		output->current_mode->flags |= WL_OUTPUT_MODE_PREFERRED;
+	base->current_mode = cur;
+	base->native_mode = cur;
 
-		/* In HiDefRemoteApp mode, free previous current_mode,
-		   since it only want to expose current mode to app */
-		wl_list_remove(&previous_mode->link);
```

Comment:
As I replied in email, we still need this, so "available mode" will be only current mode.

Reply:
Yes, your'e right - I've added a new commit to address this in a slightly different way.